text/stateful-resources-one-pager.md

```diff
@@ -0,0 +1,258 @@
+
+An additional feature we can provide is to detect when a deployment will change the policy of stateful resources. i.e:
+
+```console
+The removal policy of the following stateful resources will change:
+
+ - MyBucket (RETAIN -> DESTROY)
+ - MyDynamoTable (RETAIN -> DESTROY)
+
+Do you wish to continue (y/n)?
+```
+
+Blocking these types of changes will keep CloudFormation templates in a safe state.
+Otherwise, if such changes are allowed to be deployed, out of band operations
+via CloudFormation will still pose a risk.
+
+## High Level Design
+
+In order to provide any of the features described above, the core framework must be able
+to differentiate between stateful and stateless resources.
+
+> **Side note**: Explicitly differentiating stateful resources from stateless ones is a
+> common practice in infrastructure modeling. For example, Kubernetes
+> uses the [`StatefulSet`](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
+> resource for this.
+> And terraform has the [`prevent_destroy`](https://www.terraform.io/language/meta-arguments/lifecycle#prevent_destroy) property to mark stateful resources.
+
+### API
+
+We'd like to extend our API so it enforces every resource marks itself as either
+stateful or stateless. To that end, we propose adding the following properties to
+the `CfnResourceProps` interface:
+
+```ts
+export interface CfnResourceProps {
+  /**
+   * Whether or not this resource is stateful.
+   */
+  readonly stateful: boolean;
+
+  /**
+   * Which properties cause replacement of this resource.
+   * Required if the resource is stateful, encouraged otherwise.
+   *
+   * @default undefined
+   */
+  readonly replacedIf?: string[];
+}
+```
+
+As well as corresponding getters:
+
+```ts
+/**
+ * Represents a CloudFormation resource.
+ */
+export class CfnResource extends CfnRefElement {
+
+  public get stateful: boolean;
+
+  public get replacedIf?: string[];
+
+  constructor(scope: Construct, id: string, props: CfnResourceProps) {
+    super(scope, id);
+    this.stateful = props.stateful;
+    this.replacedIf = props.replacedIf;
+  }
+}
+```
+
+### CodeGen
+
+Adding these properties will break the generated L1 resources. We need to add some
+logic into `cfn2ts` so that it generates constructors that pass the appropriate values.
+For that, we manually curate a list of stateful resources and check that into our source code, i.e:
+
++ stateful-resources.json
+
+> The key is the resource type, and the value is the list of properties that require replacement.
+
+```json
+{
+  "AWS::S3::Bucket": ["BucketName"],
+  "AWS::DynamoDB::Table": ["TableName"],
+}
+```
+
+Using this file, we can generate the correct constructor, for example:
+
+```ts
+export class CfnBucket extends cdk.CfnResource implements cdk.IInspectable {
+
+  constructor(scope: cdk.Construct, id: string, props: CfnBucketProps = {}) {
+      super(scope, id, {
+        stateful: true,
```

Comment:
If we go the "stack policies" route, we need a way to turn off the protection, in a way that is user controllable (when they finally do want to get rid of a stateful resource).

Maybe it's a sort of `removalPolicy` though:

* `RemovalPolicy.DESTROY`
* `RemovalPolicy.RETAIN`
* `RemovalPolicy.PROTECTED`

When you want to get rid of it, you switch to `RemovalPolicy.DESTROY` first, then remove it.